memory/buffer.py

```diff
@@ -0,0 +1,244 @@
+            from utils.kitchen_dataset import get_dataset
+            dataset = get_dataset(argparse.Namespace(
+                task=self.task_name,
+                policy=policy_path,
+                data_dir=data_dir,
+            ), verbose=self.verbose)
+        elif self.domain_name == 'metaworld':
+            from utils.metaworld_dataset import get_dataset
+            dataset = get_dataset(argparse.Namespace(
+                task='pick-place',
+                policy=policy_path,
+                data_dir=data_dir,
+                noisy=True,
+            ), verbose=self.verbose)
+        else:
+            raise NotImplementedError
+
+        self.dataset = self._preprocess(dataset)
+
+        if self.verbose:
+            print('[B]', 'Dataset loaded. size: {}'.format(len(self.dataset['observations'])))
+
+    def load_expert_demos(self, data_dir='./data', policy_path=None, num_demos=10):
+        if self.domain_name == 'antmaze':
+            from utils.antmaze_dataset import get_dataset
+            dataset = get_dataset(argparse.Namespace(
+                task=self.task_name,
+                policy=policy_path,
+                data_dir=data_dir,
+                multi_start=True,
+                multigoal=True,
+                noisy=False,
+                expert_demos=num_demos,
+            ), verbose=self.verbose)
+        else:
+            raise NotImplementedError
+
+        self.dataset = self._preprocess(dataset)
+
+        if self.verbose:
+            print('[B]', 'Expert demos: {}'.format(len(self.dataset['observations'])))
+
+    def _preprocess(self, dataset):
+        if 'successes' not in dataset:
+            if 'timeouts' in dataset:
+                dataset['successes'] = np.logical_and(
+                    dataset['terminals'],
+                    np.logical_not(dataset['timeouts'])
+                )
+            elif 'rewards_sparse' in dataset:
+                dataset['successes'] = np.logical_and(
+                    dataset['terminals'],
+                    dataset['rewards_sparse'] > 0
+                )
+            else:
+                dataset['successes'] = dataset['terminals']
+
+        if self.normalize:
+            if self.obs_mean is None:
+                obs = dataset['observations']
+                self.obs_mean = obs.mean(0)
+                self.obs_std = obs.std(0)
+                self.obs_std[self.obs_std < 1e-6] = 1e-6
+            dataset['observations'] = self.normalize_observation(dataset['observations'])
+            dataset['next_observations'] = self.normalize_observation(dataset['next_observations'])
+            if self.verbose:
+                print('[B]', 'Using normalized observations')
+
+        if self.sparse_reward and 'rewards_sparse' in dataset:
+            dataset['rewards'] = dataset['rewards_sparse']
+            if self.verbose:
+                print('[B]', 'Using sparse reward')
+
+        return dataset
+
+
+class SubtrajBuffer(Buffer):
```

Comment:
Could you add a comment or a string to explain the difference between this class and the `Buffer` class?

Reply:
This class returns sub-trajectories instead of individual samples as items (note that `subtraj_slice` is a list of indices in `__getitem__`). I'll add comments in the class.